### ISSUE TITLE
Implement player elimination

### ISSUE THREAD

**Issue by Dode123** [Longwelwind/swords-and-ravens#576]
Albeit the rules say that no player is ever eliminated, this doesn't work well for online versions of the game. There is no reason for a player to be expected to stick around once he has no castles or land units left, as he can no longer get any VPs, the goal of the game. 
All it does, is make is much more likely to get stalled games as the players leave, especially during a live game. You cannot expect a person to sit around for another 3 hours if he has nothing left.
Player elimination is also present on Thronemaster. 
Elimination works very similar to vassals across the board. This will also make it much easier to implement. 

**Once a player:**
- Does not have any castles.
- Does not have any land units.
- Does not hold a dominance token. (if he does, this check will be performed again on the next round). 

**Eliminated from the game at the end of the current game round.** 

**Ship units:** 
- Using MoD/Vassals: Convert any remaining ship units (if any) to vassal units of his house. 
- Using base game/Not vassals: Convert any remaining ship units (if any) to neutral forces (functions just like a neutral garrison). 

**Wildings and influence tracks:** 
- During bids, the eliminated house automatically bids 0, regardless of how many power tokens he had left.
- The eliminated house cannot be chosen for wildling effects (immune, just like vassals). 
- The eliminated house can be moved by the Iron Throne holder, and if he holds a dominance token, its effect will go to the second house on the track (just like vassals).

**Comment by gereon77** [comment]
I am not really a fan of this idea because a user can still influence a game with his ships and power tokens.
But yea, atm it is frustrating when you have no units and castles left and you have to stay in a live game (i see no problem for PBEM games) just for biddings or sea battles.
So I‘d vote to implement #139  so you see if a user left and if so, just convert him to a vassal, which in fact is a proper elimination system.

**Comment by Dode123** [comment]
> I am not really a fan of this idea because a user can still influence a game with his ships and power tokens.
> But yea, atm it is frustrating when you have no units and castles left and you have to stay in a live game (i see no problem for PBEM games) just for biddings or sea battles.
> So I‘d vote to implement #139 so you see if a user left and if so, just convert him to a vassal, which in fact is a proper elimination system.

Yes, this is for live games. 
Just changing to vassals will allow other players to move them up on the tracks to a dominance token. How is that fair? Easier to just remove them from the equation IMO.
Also, you will have games without MOD/vassals, what then? This solution also fixes that.

**Comment by gereon77** [comment]
> Just changing to vassals will allow other players to move them up on the tracks to a dominance token.

A player being converted into a vassal isn’t immediately moved to the bottom of all tracks, is he? So if the eliminated player had a dominance token it will be handed over to the second position on the track... that’s how it goes when a player is turned into a vassal. I see no problem by that.

But yea, you are right with the base game thing. It’s a bit more intricate here. Maybe vassals could be an option which can be activated for a running game to replace a player (which would not work for 3p base games but I guess when someone leaves here it has to be aborted anyways).

**Comment by Dode123** [comment]
> > Just changing to vassals will allow other players to move them up on the tracks to a dominance token.
> 
> A player being converted into a vassal isn’t immediately moved to the bottom of all tracks, is he? So if the eliminated player had a dominance token it will be handed over to the second position on the track... that’s how it goes when a player is turned into a vassal. I see no problem here.
> 
> But yea, you are right with the base game thing. It’s a bit more intricate here. Maybe vassals could be an option which can be activated for a running game to replace a player (which would not work for 3p base games but I guess when someone leaves here it has to be aborted anyways).

Sure, but what if you play without vassals in a strict base game? Two things about that, that I don't like. 
1) Influence shifting: Suddenly giving the blade to another house, can have a big impact. If you can't do both marches for 5-7 VPs as intended, as your opponent can win one of the battles with blade. 
I feel like that is a bit.. Unfortunate. Since the game was hosted and played for all but the last round, without vassals and its mechanics, and instead introduced during a critical moment of the 10th round. That's not fair. 
2) Adding vassals means someone can suddenly control them! In addition, adding their special tokens, you can battle with them (to refresh your house cards), they can Jaquen you (can kill your end game plan if he picks the wrong card), and most importantly, another player will gain control of those remaining ship units. This could also be huge for the end result of the game, if it is during round 9-10. 

So that is why I DO NOT want vassal system in games that have no MoD elements in them (without a option). We need something that works well for the base game as well. The purposed system is very similar to the vassal system, but with valid core differences for minimal impact. 

So we could have the players decide for themselves what they want to do. 
**MoD/vassal game**: Vassals as standard, cannot be changed. (as per official MoD rules). 
**Base game/no vassals**: The below garrison system as standard, but it can be changed to vassal system for an eliminated house, if **ALL** players vote for it. 

So TLDR, just two core differences: 
- Ships are **neutral garrisons**, not units that you can battle with to shuffle cards. For minimal impact on the other houses. 
- Ships have no "commander", you cannot place orders, win combats and gain PTs with them, support yourself etc. They are just static.

**Comment by gereon77** [comment]
So this means you can't be eliminated when you hold a dominance token? Unlike it's done on other platforms where an eliminated player is moved to the bottom on all tracks to avoid stalled games due to Westeros decisions...
Well that sounds like a possible way to go. I still just don't like player elimination at all but yea... For an online live game we probably need this.

**Comment by Dode123** [comment]
> So this means you can't be eliminated when you hold a dominance token? Unlike it's done on other platforms where an eliminated player is moved to the bottom on all tracks to avoid stalled games due to Westeros decisions...
> Well that sounds like a possible way to go. I still just don't like player elimination at all but yea... For an online live game we probably need this.

Exactly. Yeah I think so.. For PBEM of course this isn't a issue tbh.

Most will still play with vassals though, which will just keep the "official" version with adding a vassal  instead.

**Comment by Dode123** [comment]
Up for grab? @Longwelwind

**Comment by Longwelwind** [comment]
It's a really big feature that requires modifications in a lot of parts of the codebase, since all possible game states need to have automated actions, in case an eliminated player need to do something. This would complexify a lot the codebase. That's why I'd prefer to first have the vassal feature (and proper player replacement by a vassal) before implementing this.

I understand the concerns about the possible impact on gameplay due to players replacing someone by a vassal (influence shifting, ...), but from a development perspective, I'd rather see if replacement by a vassal is enough to solve the issue before going to a more complex solution.

**Comment by gereon77** [comment]
@Longwelwind : You know I am not really a fan of player elimination, especially when vassals can jump in the future. I just wondered why it would be this difficult to do it? Why every GameState would need automated actions? When an player is considered as eliminated he should never become "active" again. Instead of doing automated actions I would try to solve it by never let him become active again by removing the eliminated player from all possible arrays. `ingame.players`, `game.houses`, `game.tracks`, `etc.` and add an extra handling in `ResolveSingleMarchOrderGameState` and treat an attack against an eliminated house like an attack against a neutral force. Would that be possible or do I miss sth which would make this more difficult than I think. Right now that seems doable instead of adding automated actions to all possible game states. And please don't understand me wrong, I still vote for waiting for vassals instead of doing this. I am just interested if my idea technically would work.

**Comment by Longwelwind** [comment]
You can't remove the house in `game.houses` because then game logs, for example, would be un-serializable. It would not be that simple to treat `ResolveSingleMarchOrderGameState` as a neutral attack as neutral attacks are not handled during a combat.

One way to do this would be to have a property `Player.eliminated` to mark them as eliminated.

All in all, it would greatly increase the complexity of the code, I think. I would rather have an unified way of handling player elimination (with Vassals) than spending time implementing this, and then having to maintain it after.
